+sw_tests/+unit_tests/unittest_spinw_addatom.m

```diff
@@ -0,0 +1,185 @@
+            testCase.swobj.addatom('r', pos_vector)
+            testCase.assertEqual(testCase.swobj.unit_cell, ...
+                testCase.default_unit_cell);
+        end
+        
+        function test_add_multiple_atom_with_only_position(testCase)
+            pos = [[0; 0; 0] [0; 0; 0.5]];
+            testCase.swobj.addatom('r', pos)
+            unit_cell = testCase.swobj.unit_cell;
+            testCase.assertEqual(unit_cell.r, pos)
+            testCase.assertEqual(unit_cell.S, [0,0])  % default non-mag
+            testCase.assertEqual(unit_cell.label, {'atom_1', 'atom_2'})
+        end
+
+        
+        function test_add_atom_with_update_true_different_spin(testCase)
+            pos = [0; 0; 0];
+            testCase.swobj.addatom('r', pos, 'label', 'atom1')
+            testCase.swobj.addatom('r', pos, 'S', 1, 'label', 'atom1')
+            unit_cell = testCase.swobj.unit_cell;
+            testCase.assertEqual(size(unit_cell.r), [3, 1])  % 1 atom
+            testCase.assertEqual(unit_cell.S, 1)  % updated spin
+            testCase.assertEqual(unit_cell.label, {'atom1'})
+        end
+        
+        function test_add_atom_update_false_different_spin(testCase)
+            pos = [0; 0; 0];
+            testCase.swobj.addatom('r', pos, 'label', 'atom1')
+            testCase.verifyWarning(...
+                @() testCase.swobj.addatom('r', pos, 'S', 1, ...
+                    'label', 'atom1', 'update', false), ...
+                'spinw:addatom:WrongInput')  % warns occ > 1
+            unit_cell = testCase.swobj.unit_cell;
+            testCase.assertEqual(size(unit_cell.r), [3, 2])  % 2 atoms
+            testCase.assertEqual(unit_cell.S, [0, 1])
+        end
+        
+        function test_add_atom_will_not_update_different_pos(testCase)
+            pos = [0; 0; 0];
+            testCase.swobj.addatom('r', pos, 'label', 'atom1')
+            testCase.swobj.addatom('r', pos + 0.5, 'S', 1, 'label', 'atom1')
+            unit_cell = testCase.swobj.unit_cell;
+            testCase.assertEqual(size(unit_cell.r), [3, 2])  % 2 atom
+            testCase.assertEqual(unit_cell.S, [0, 1])
+        end
+        
+        function test_add_atom_named_ion_lookup(testCase)
+            testCase.swobj.addatom('r', [0; 0; 0], 'label', 'Mn3+')
+            unit_cell = testCase.swobj.unit_cell;
+            testCase.assertEqual(size(unit_cell.r), [3, 1])
+            testCase.assertEqual(unit_cell.S, 2)
+            testCase.assertEqual(unit_cell.ox, 3)
+            testCase.assertEqual(unit_cell.Z, int32(25))
+            testCase.assertEqual(unit_cell.b(1), -3.73) % coh scat. len
+            testCase.assertEqual(unit_cell.ff(:,1), [0.4198; 6.926972])
+            testCase.assertEqual(unit_cell.label, {'Mn3+'})
+            % check default values
+            for field = {'occ', 'A', 'biso'}
+                testCase.assertEqual(unit_cell.(field{:}), ...
+                    testCase.default_unit_cell.(field{:}))
+            end
```

Comment:
This validation could be reduced to one function and called like `testCase.verify_unit_cell({'r', [0; 0; 0;], 'S', 2}...` where the input are the names/values of the non-default attributes and it automatically checks the default attributes, it would avoid a lot of duplication across different tests, and make sure we consistently check everything.

The same for `addg` and `addaniso` too (and even `addmatrix` e.g. we can't test the default `color` but we can test it is the expected type/size)

Reply:
Thanks, this is a good idea for some `addatom` tests (although in the particular case above I don't think it will work because  there isn't 1:1 relation between properties set in the function arguments and non-default fields in the relevant struct e.g. when setting a label corresponding to a valid ion name `Fe3+` it will set non-default values for `b`, `ff`, `Z`). If you are concerned about the loop on L134, I could change `verify_obj` in the unit test super class to accept a cell of field names?

For the other functions I don't think such a function would actually be very helpful/work:
- For `addg` and `addansio` the input arguments don't directly relate to the struct fields at all, e.g. `addaniso('A1')` could result in ` single_ion.aniso = [1,1]` or `1` depending on the number of magnetic ions in the unit cell and `addaniso('A1', 'atom_2')` results in `single_ion.aniso = [0,1]`  (the other tests are more complicated with multiple calls to `addaniso` anyway).
- For `addmatrix` the object to test `matrix.mat` depends on the form of the `value` passed: a 3x3 matrix is appended, a single value is interpreted as a diagonal matrix etc. A lot of that logic would have to be reproduced in the helper function.

Reply:
I think there might be a bit of a misunderstanding here, the struct arguments we're checking don't have to relate to the input arguments, because like what you've already written you're manually checking that `S = 2` and `ox = 3`, its just putting all those `assertEqual` statements into another function, this means you don't have to write out `assertEqual` each time, and it ensures you always do the same checks and don't miss out any attributes. e.g.
you just call `testCase.verify_unit_cell(unit_cell, struct('S', 2, 'ox', 3...` (sorry I missed out the unit_cell argument in my previous example) which could do something like:
```
function verify_unit_cell(testCase, unit_cell, expected_unit_cell)
    default_unit_cell = struct({'r', [0; 0; 0;], 'S' 0, 'ox', 0...})
    % merge expected and default unit cell, doing this manually because matlab is weird, there may be a better way
    for attr = fieldnames(default_unit_cell)
        if ~isfield(expected_unit_cell)
            expected_unit_cell.(attr) = default_unit_cell.(attr)
        end
    end
    % Now test fields
    testCase.assertEqual(unit_cell.r, expected_unit_cell.r)
    testCase.assertEqual(unit_cell.S, expected_unit_cell.S)
    testCase.assertEqual(unit_cell.S, expected_unit_cell.ox)...
```
Does this help? We can have a zoom meeting if we still misunderstand each other (maybe there's something I've missed?)

Reply:
Ah I see, yes I had misunderstood, sorry! I actually went about it a bit differently (see commit https://github.com/SpinW/spinw/pull/85/commits/876789def3e20f53b83294f1c98d9fadcc074fcb)
I used the parameterised testing in MATLAB to do a loop over name,value pairs - that way I didn't have to write out multiple asserts (you could do a similar loop in your function above). But as I have written it it's only relevant for input arguments that change the struct field with the same name (this wasn't the case for any other tests hence I didn't define a separate helper function).

I could use the above to test the atom lookup functionality (2 tests) and I can see that this would mean every field of `unit_cell` is guaranteed to be checked so i will change and force-push.

Reply:
OK do you want this `verify_unit_cell` in the unit test super class then rather than in the addatom test file?

Reply:
Would something like this fit the bill?
```
function verify_struct(testCase, default_struct, actual_struct, non_default_fields_struct)
    expected_struct = default_struct;  % copies!
    for field = fieldnames(non_default_fields_struct)
        expected_struct.(field) = non_default_fields_struct.(field);
    end
    testCase.assertEqual(actual_struct, expected_struct)
end
```